taf/testlib/linux/collectd.py

```diff
@@ -31,163 +31,151 @@
-           'change_param': {'cmd': [r"sed -i '/^<Plugin \({plugin}\|\"{plugin}\"\)>/,/^<\/Plugin>/s/\(^\s*\)\({par}.*\)/\1{par} {val}/w /dev/stdout' {collectd_conf}"],
-                            'kwargs_required': True},
-           'change_global': {'cmd': [r"sed -i '/^<LoadPlugin \({plugin}\|\"{plugin}\"\)>/,/^<\/LoadPlugin>/s/\(^\s*\)\({par}.*\)/\1{par} {val}/w /dev/stdout' {collectd_conf}"],
-                             'kwargs_required': True},
-           'insert_param': {'cmd': [r"sed -i '/^<Plugin \({plugin}\|\"{plugin}\"\)>/a\\t{par} {val}' {collectd_conf}"],
-                            'kwargs_required': True},
-           'disable': {'cmd': [r"sed  -i '/^LoadPlugin {plugin}/s/^[^#]/#&/w /dev/stdout' {collectd_conf}",
-                               r"sed  -i '/^<Plugin \({plugin}\|\"{plugin}\"\)>/,/^<\/Plugin>/s/^[^#]/#&/w /dev/stdout' {collectd_conf}"],
-                       'kwargs_required': False},
-           'disable_global': {'cmd': [r"sed  -i '/^<LoadPlugin \({plugin}\|\"{plugin}\"\)>/,/^<\/LoadPlugin>/s/^[^#]/#&/w /dev/stdout' {collectd_conf}"],
-                              'kwargs_required': False},
-           }
-
-
-class CollectdConfCommandGenerator(object):
-    def __init__(self, action, command_generator, collectd_conf, plugin_list=PLUGINS):
-        """
-        @brief  Initialize collectd conf command generator class
-        """
-        super(CollectdConfCommandGenerator, self).__init__()
-        self.plugin_list = plugin_list
-        for plugin in self.plugin_list:
-            setattr(self, plugin, command_generator(action, plugin, collectd_conf))
-
-
-class CollectdPluginsManager(object):
-    def __init__(self, action, collectd_conf_commands, run):
-        """
-        @brief  Initialize collectd conf manager class
-        """
-        super(CollectdPluginsManager, self).__init__()
-        for cmd in collectd_conf_commands.plugin_list:
-            setattr(self, cmd,
-                    self.generate_run_function(run, getattr(collectd_conf_commands, cmd), action))
 
-    @staticmethod
-    def generate_run_function(run_func, command, action):
-        def run(**kwargs):
-            return run_func(command(**kwargs))
-        return run
+Restart collectd service
+env.lhost[1].ui.collectd.restart()
+"""
+import collections
 
+from testlib.linux import service_lib
+from testlib.custom_exceptions import CustomException
 
-def collectd_conf_action(action, run_func, collectd_conf):
-    collectd_conf_commands = CollectdConfCommandGenerator(action, command_generator, collectd_conf, PLUGINS)
-    return CollectdPluginsManager(action, collectd_conf_commands, run_func)
+INDENT = ' ' * 4
+PARAM_BOILERPLATE = '{indent}{name} {value}\n'
+TAGGED_BOILERPLATE = ('{indent}<{tag}{space}{name}>\n'
+                      '{params}'
+                      '{indent}</{tag}>\n')
 
 
-def command_generator(action, plugin, collectd_conf):
+def build_tagged_section(config_data, indent_level=-1, res=''):
```

Comment:
I've checked suggested change and in progress of applying it.
Though I doubt that `IntentationContext` will serve properly.
The side effect of it seems to be initial indent of 4 whitespaces, while no indentation is expected for top level tag.
Currently I don't see any elegant way out to fix it.
So I wonder if I miss it, or I may merge indentation handling into TagContext logics. 
Please let me know your opinion.

Current context managers I come with, based on your suggestion:
```
def build_tagged_section(config_data, indent=None, buffer=None):
    # ... skipped docstrings ...
    if not buffer:
        buffer = StringIO()
    if not indent:
        indent = IndentationContext(buffer=buffer)
    for k, v in config_data.items():
        with indent:
            if isinstance(v, collections.Mapping):
                with TagContext(buffer, indent, *k.split(' ', 1)):
                    build_tagged_section(v, indent, buffer)
            elif not isinstance(v, str) and isinstance(v, collections.Iterable):
                buffer.write(''.join(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=x) for x in v))
            else:
                buffer.write(PARAM_BOILERPLATE.format(indent=indent.pad, name=k, value=v))
    return buffer.getvalue()


class IndentationContext(object):
    def __init__(self, char=' ', count=4, buffer=None):
        super().__init__()
        self._char = char
        self._count = count
        self._pad = ''
        self._buffer = buffer

    @property
    def pad(self):
        return self._pad

    def __enter__(self):
        self._pad += self._char * self._count

    def __exit__(self, *args, **kwargs):
        self._pad = self._pad[:-self._count]


class TagContext(object):
    def __init__(self, buffer, indent, *tag_args):
        super().__init__()
        self._buffer = buffer
        self._indent = indent
        assert tag_args
        self._tag_args = tag_args

    def __enter__(self):
        self._buffer.write(self._indent.pad)
        self._buffer.write('<')
        self._buffer.write(' '.join(self._tag_args))
        self._buffer.write('>\n')

    def __exit__(self, *args, **kwargs):
        self._buffer.write(self._indent.pad)
        self._buffer.write('</')
        self._buffer.write(self._tag_args[0])
        self._buffer.write('>\n')
```
Please note that `__enter__()` and `__exit__()` methods of TagContext were updated with newline character after closing tag bracket to enforce multiline text that was missing in suggested "sample" code.